scripts/benchmark.sh

```diff
@@ -0,0 +1,88 @@
+#!/usr/bin/env bash
+# Benchmark pipes.sh by revisions
+# Copyright (c) 2018 Pipeseroni/pipes.sh contributors
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+# Usage: scripts/benchmark.sh <revision>...
+#
+# Example:
+#
+#     $ scripts/benchmark.sh f7d0941 $(git tag) HEAD
+#     f7d0941   :    564 c/s
+#     2013-02-01:    534 c/s
+#     2013-02-06:    534 c/s
+#     [snip]
+#     v1.2.0    :    529 c/s
+#     v1.3.0    :    492 c/s
+#     HEAD      :   3030 c/s
+
+
+PIPESSH='pipes.sh'
+LIMIT=1000
+
+SED_E=(
+    -e 's/read\|sleep\|cat/:/g'  # NOP
+    -e 's/! :/:/'                # ! read -> ! : -> :
+)
+SED_CLNUP=("${SED_E[@]}" -e 's/&& t=0/\&\& cleanup/')
+SED_BREAK=("${SED_E[@]}" -e 's/&& t=0/\&\& break/')
+
+
+# monkey-patch gitrevision $1 pipes.sh:
+# 1. make read/sleep/cat nop
+# 2. exit when reach -r LIMIT
+#
+# Some early commits use sleep to work around Bash < 4 for delay < 1 second,
+# and cat is used to clear out standard input, which could hang if not made
+# nop.
+mkp_pipes() {
+    local GIT=(git cat-file --textconv "$1:$PIPESSH")
+    if "${GIT[@]}" | grep cleanup &>/dev/null; then
+        "${GIT[@]}" | sed "${SED_CLNUP[@]}"
+    else
+        "${GIT[@]}" | sed "${SED_BREAK[@]}"
+    fi
+}
+
+
+# time the monkey-patched gitrevision $1 pipes.sh with Bash builtiin time.
+# The stdout and stderr are swapped.  So the pipes are on stderr and time's
+# result (real %e) is on stdout.
+run() {
+    TIME=%e time bash <(mkp_pipes "$1") -r $LIMIT 3>&2 2>&1 1>&3
+}
+
+
+# benchmark a gitrevision ($1) pipes.sh, the result is pipe chars per second.
+benchmark() {
+    local t=$(run "$1" 2>/dev/null)
```

Comment:
I think redirecting the output to `/dev/null` hides a bottleneck. Before `c582ec0`, the bottleneck was calling `tput`, but after that the bottleneck becomes output to the terminal. This is more obvious with a higher `LIMIT`. Setting `LIMIT=10000` and running `./scripts/benchmark.sh HEAD`, I get 8382 c/s when writing to `/dev/null` and 7163 c/s writing to my terminal. I repeated the measurements a few times and these seem quite consistent.

Reply:
Actually, I did that on purpose.  Please allow me to use PR (public relation) / marketing tactics:

> We do not hide the bottleneck, it's *excluded* from the benchmark as our users would not care for them.

Anyway, the real rationale is explained in new `scripts/README`, but I also added `TIMEF` (along with `LIMIT`) environment variable to allow the switch.

From the beginning, I know terminal render is a big black box(es), and after added `TIMEF=time_disp`, I see how big it is, and even more certain that we need to exclude it.

For instance, on this old computer, the time for that `HEAD` could vary from 1600s to 3000s c/s.  There is no way that we can relies on that with render time included.  Because as each revision is benchmarked, they could subject to that level of variation, and you can't tell the levels of each benchmark is affecting.  Not only we can't compare across revisions, but we can't even compare two benchmark runs on the same revision, simply because the level of variation is just *ridiculous*.

However, with `TIMEF=time_ndis`, although I am still seeing about a couple of hundred c/s fluctuation, with increase of `LIMIT`, that seems to reduce or converge to a point.  (The default 1000 might be too small, but with that value `time_disp` still gives me big variation)

Granted, terminal render plays big part of performance, but it's not a performance that we can directly improve and gauge accurately.  And as I have used the phrase "black box," what you think you have improved might have opposite effect in other environments.

Therefore, the only thing for certain is the internal Bash performance, which I would believe virtually whatever environment the benchmark script is run with -- terminal emulator or console, X Window or Wayland -- we can always get a consistent performance index on the same machine across those environment with same Bash version, because we simply doesn't include those in the index.

I don't know if my reasoning sound enough, since we can switch with `TIMEF=time_disp`, all should be covered.

----

(I wasn't sure what to call *the output to terminal*, so I initially used the term "terminal displaying", then switched to "terminal render," but forgot to rename the timing helper functions.)

Reply:
Sidenote: I also thought about writing a gnuplot, but somehow, I didn't do it at original push, probably the textual output was enough.

I was considering to add a plot script after reading your comment, because it might be helpful to see how much performance drop in terminal render, but after seeing the huge variation, nope. That's never going to be meaningful.

Reply:
Makes sense to me :+1: 

It would be useful to benchmark terminal performance if you were a masochist and decided to reimplement the "cursor optimisation" part of curses in bash, but that would be madness. I don't even know if local cursor movements are faster than non-local cursor movements in a terminal emulator. Hmm... I should benchmark `cpipes` linked to ncurses libraries compiled with and without cursor optimisation.

Reply:
Regarding plots, I've been [toying with][1] the idea of automatically generating gifs or screenshots of `cpipes` for programmatically updating `README.md`. Including a plot of performance improvements in `README.md` that is automatically updated each release would be cool.

I wonder if the performance drop due to rendering can be reduced by sorting the list of characters to be written by their *(x, y)* coordinates so that the distance the cursor has to jump is minimised. I've got no idea if that actually affects terminal emulators, though.

[1]: https://gist.github.com/StefansM/813a319543c8196f48a6ce8dd144cb15

Reply:
I dare you to reinvent the wheel!  /jk

From my experience, if the *optimization* isn't cutting the lines shorter, reducing the number the lines, or stripping variables; instead, more control flows and/or variables, likely the performance would get worse.  If "cursor optimization" is re-implemented in pure Bash, I would bet its performance is going to be horrible.  Sorting in Bash is getting me severe chill.

(Whispering Bash builtin, or hacking that part of ncurses into builtin, or do the unimaginable because you can `enable bash-ncursesw`)

I try to remind myself every time I write in Bash, keep it stupidly simple and *flat*, don't do anything fancy (such as indirect expansion).  If it's fairly human-readable, it'd be fine.  Don't use external command if you can do it in Bash with a line or two, especially when in loops.  Whenever I see someone put a long pipeline with several commands in a loop like `grep` and `cut`, which can be done with one `sed`, my heart bleeds a bit.

The overhead of spawning new process is really expensive, which was the main reason for that commit.  When I finally noticed that `tput cup` line, I immediately replaced it with manual escape code print-out.  I didn't even know how much the performance was improved until yesterday, although `tput` might account the most of the performance cost.

----

That's a very cool script and cool idea!  Perhaps license it and generalize it?

I've never thought of using xvfb for recording, although I have heard of it, but never even read its full name, wasn't aware what its purpose is.  Pretty sure I thought it was something can be used with MPlayer's `-vo`. :man_facepalming: 

By the way, GitHub's render box's 888px in width last time I checked.  A full edge-to-edge image/GIF would look very nice in the README.